### PR TITLE
Clean up nonlinear extension

### PR DESCRIPTION
**Pull request by nafur**
This PR performs some cleanups in the nonlinear extension, in particular it removes the old lemma collection scheme. It is no longer needed, as all subsolvers use the inference manager.

### CODE REVIEW

**Comment by nafur** [comment]
Benchmark results:
This PR has no effect with `--no-nl-ext`, but improves performance across the board with `--nl-ext`.

File: src/theory/arith/nl/nonlinear_extension.cpp [hunk 1]
```diff
@@ -406,17 +405,13 @@ int NonlinearExtension::checkLastCall(const std::vector<Node>& assertions,
...
-  }
 
   // init last call with IAND
   d_iandSlv.initLastCall(assertions, false_asserts, xts);
 
-  if (d_im.hasUsed() || !lems.empty())
+  if (d_im.hasUsed())
   {
-    unsigned count = lems.size() + d_im.numPendingLemmas() + d_im.numSentLemmas();
+    unsigned count = d_im.numPendingLemmas() + d_im.numSentLemmas();
```
Comment (ajreynol) [inline src/theory/arith/nl/nonlinear_extension.cpp:414]:
```suggestion
    size_t count = d_im.numPendingLemmas() + d_im.numSentLemmas();
```

File: src/theory/arith/nl/nonlinear_extension.cpp [hunk 2]
```diff
@@ -442,7 +437,7 @@ int NonlinearExtension::checkLastCall(const std::vector<Node>& assertions,
     d_trSlv.checkTranscendentalInitialRefine();
     if (d_im.hasUsed())
     {
-      unsigned count = lems.size() + d_im.numPendingLemmas() + d_im.numSentLemmas();
+      unsigned count = d_im.numPendingLemmas() + d_im.numSentLemmas();
```
Comment (ajreynol) [inline src/theory/arith/nl/nonlinear_extension.cpp:440]:
```suggestion
      size_t count = d_im.numPendingLemmas() + d_im.numSentLemmas();
```

File: src/theory/arith/nl/nonlinear_extension.cpp [hunk 3]
```diff
@@ -473,7 +468,7 @@ int NonlinearExtension::checkLastCall(const std::vector<Node>& assertions,
     d_trSlv.checkTranscendentalMonotonic();
     if (d_im.hasUsed())
     {
-      unsigned count = lems.size() + d_im.numPendingLemmas() + d_im.numSentLemmas();
+      unsigned count = d_im.numPendingLemmas() + d_im.numSentLemmas();
```
Comment (ajreynol) [inline src/theory/arith/nl/nonlinear_extension.cpp:471]:
```suggestion
      size_t count = d_im.numPendingLemmas() + d_im.numSentLemmas();
```

File: src/theory/arith/nl/nonlinear_extension.h [hunk 4]
```diff
@@ -151,7 +146,7 @@ class NonlinearExtension
    * may have information regarding how to construct a model, in the case that
    * we determined the problem is satisfiable.
    */
-  bool modelBasedRefinement(std::vector<NlLemma>& mlems);
+  bool modelBasedRefinement();
```
Comment (ajreynol) [inline src/theory/arith/nl/nonlinear_extension.h:149]: Can you update the documentation for this method (it doesnt add lemmas to lems)?

**Review by ajreynol — CHANGES_REQUESTED** [review]
This looks good, just a few minor requests.

**Review by ajreynol — APPROVED** [review]

**Pull request merged**